src/java/htsjdk/samtools/util/AsyncBlockCompressedInputStream.java

```diff
@@ -0,0 +1,191 @@
+ * Note that this implementation is not synchronized. If multiple threads access an instance concurrently, it must be synchronized externally. 
+ */
+public class AsyncBlockCompressedInputStream extends BlockCompressedInputStream {
+    private static final int READ_AHEAD_BUFFERS = (int)Math.ceil(Defaults.NON_ZERO_BUFFER_SIZE / BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE);
+    private static final Executor threadpool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors(),new ThreadFactory() {
+            public Thread newThread(Runnable r) {
+                Thread t = Executors.defaultThreadFactory().newThread(r);
+                t.setDaemon(true);
+                return t;
+            }
+        });
+    private final BlockingQueue<DecompressedBlock> mResult = new ArrayBlockingQueue<>(READ_AHEAD_BUFFERS);
+    private final BlockingQueue<byte[]> freeBuffers = new ArrayBlockingQueue<>(READ_AHEAD_BUFFERS);
+    private final Semaphore running = new Semaphore(1);
+    private volatile boolean mAbort = false;
+    private AsyncBlockCompressedInputStreamRunnable runnable = new AsyncBlockCompressedInputStreamRunnable();
+
+    public AsyncBlockCompressedInputStream(final InputStream stream) {
+        super(stream, true);
+    }
+
+    public AsyncBlockCompressedInputStream(final File file)
+        throws IOException {
+        super(file);
+    }
+
+    public AsyncBlockCompressedInputStream(final URL url) {
+        super(url);
+    }
+
+    public AsyncBlockCompressedInputStream(final SeekableStream strm) {
+        super(strm);
+    }
+    
+    protected DecompressedBlock nextBlock(byte[] bufferAvailableForReuse) {
+        if (bufferAvailableForReuse != null) {
+            freeBuffers.offer(bufferAvailableForReuse);
+        }
+        return nextBlockSync();
+    }
+    
+    @Override
+    protected void prepareForSeek() {
+        flushReadAhead();
+        super.prepareForSeek();
+    }
+
+    @Override
+    public void close() throws IOException {
+        // Suppress interrupts while we close.
+        final boolean isInterrupted = Thread.interrupted();
+        mAbort = true;
+        flushReadAhead();
+        super.close();
+        if (isInterrupted) Thread.currentThread().interrupt();
+    }
+    /**
+     * Foreground thread blocking operation that aborts all read-ahead tasks
+     * and flushes all read-ahead results.
+     */
+    private void flushReadAhead() {
+        final boolean abortStatus = mAbort;
+        mAbort = true;
+        try {
+            // block until the thread pool operation has completed
+            running.acquire();
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interrupted waiting for decompression thread", e);
+        }
+        // flush any read-ahead results
+        mResult.clear();
+        mAbort = abortStatus;
+        running.release();
+    }
+    private void ensureReadAhead() {
+        if (running.tryAcquire()) {
```

Comment:
so tryAcquire means that  `ensureReadAhead` may sometimes not ensure that read ahead happened?